### PR TITLE
Ongoing PR

### LINKED ISSUE #2: Review Feed Screen
- [x] Show reviews
- [x] Show media info in review post
- [x] Endless Scrolling
- [x] Add media to 'my list'
- [x] Double Tap to add to 'my list'
- [x] Implement Snap Scroll

### LINKED ISSUE #4: Compose Review Screen
- [x] Research APIs
- [x] Post Screen Fragment
- [x] Push reviews to Parse
- [x] Media Search implemented with one API
- [x] Implement search utilizing the 3 APIs
- [x] Attach selected media to review

### PR DESCRIPTION
**Pull request by jackviel**
Worked on implementing search with the books, movies, and music APIs, worked on the Results Adapter, Results Model, API connectivity, and results rendering when searching in the review compose fragment with the end goal of searching all 3 APIs on the same SearchView to show books, movies, and music based on the same query.

When a query is entered in the SearchView, it'll search through each API, then it'll add the results to a results array, and then append the results to the results adapter RecyclerView.

- Tested for queries that should bring up results for only one API, to check if there aren't any results in one API, it should only add the correct results.
- Also tested for queries that gather less than the minimum amount of results displayed in the RecyclerView results.

### CODE REVIEW

**Comment by jackviel** [comment]
Worked on attaching the selected media result to the review posted.

When a result is clicked in the result adapter RecyclerView it saves the result selection so that when the submit button is pushed the result details are sent with the rest of the review details to be saved onto the parse backend as a review post class entry.

- Tested attempting to save a review post without any media selections taken.

**Comment by jackviel** [comment]
Worked on displaying the attached media on the feed review post, also added a review title field on the compose fragment screen.

- Tested trying to submit a review with an empty review title field.

Current app status:

Kapture 2022-06-28 at 11 04 19

File: app/src/main/java/com/example/captstone/ResultAdapter.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    // Define listener member variable
+    private OnItemClickListener listener;
+
+    // Define the listener interface
+    public interface OnItemClickListener {
+        void onItemClick(View itemView, int position);
+    }
+
+    // Define the method that allows the parent activity or fragment to define the listener
+    public void setOnItemClickListener(OnItemClickListener listener) {
```
Comment (brianwal) [inline app/src/main/java/com/example/captstone/ResultAdapter.java]: Is it possible to set this in the constructor?

Comment (brianwal) [reply app/src/main/java/com/example/captstone/ResultAdapter.java]: If not do you need to handle error case when already set

File: app/src/main/java/com/example/captstone/ResultAdapter.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    @Override
+    public ResultAdapter.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
+        Context context = parent.getContext();
+        LayoutInflater inflater = LayoutInflater.from(context);
+
+        // Inflate the custom layout
+        View resultView = inflater.inflate(R.layout.item_result, parent, false);
+
+        // Return a new holder instance
+        ResultAdapter.ViewHolder viewHolder = new ResultAdapter.ViewHolder(resultView, listener);
```
Comment (brianwal) [inline app/src/main/java/com/example/captstone/ResultAdapter.java]: Do you need to handle listener == null case?

File: app/src/main/java/com/example/captstone/fragments/ComposeFragment.java [hunk 3]
```diff
@@ -5,29 +5,52 @@
...
+import okhttp3.Response;
+
 public class ComposeFragment extends Fragment {
 
     public static final String TAG = "ComposeFragment";
 
     private EditText etBody;
     private Button bSubmit;
     public ProgressBar pb;
+    public SearchView svMedia;
```
Comment (brianwal) [inline app/src/main/java/com/example/captstone/fragments/ComposeFragment.java]: No using consistant style in code

File: app/src/main/java/com/example/captstone/fragments/ComposeFragment.java [hunk 4]
```diff
@@ -43,6 +66,74 @@ public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceStat
...
+
+        aResults = new ArrayList<>();
+        resultAdapter = new ResultAdapter(getContext(), aResults);
+
+        resultAdapter.setOnItemClickListener(new ResultAdapter.OnItemClickListener() {
+            @Override
+            public void onItemClick(View itemView, int position) {
+                Toast.makeText(
+                        getContext(),
+                        "An item at position " + position + " clicked!",
```
Comment (brianwal) [inline app/src/main/java/com/example/captstone/fragments/ComposeFragment.java]: String format over multiple concat

File: app/src/main/java/com/example/captstone/fragments/ComposeFragment.java [hunk 5]
```diff
@@ -43,6 +66,74 @@ public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceStat
...
+                                aResults.clear();
+                                // Load model objects into the adapter
+                                for (Result result : results) {
+                                    aResults.add(result); // add result through the adapter
+                                }
+                                resultAdapter.notifyDataSetChanged();
+                            }
+                        } catch (JSONException e) {
+                            // Invalid JSON format, show appropriate error.
+                            e.printStackTrace();
```
Comment (brianwal) [inline app/src/main/java/com/example/captstone/fragments/ComposeFragment.java]: poor error handling

File: app/src/main/java/com/example/captstone/models/Result.java [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+import android.text.TextUtils;
+
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+import java.util.ArrayList;
+
+public class Result {
+    private String openLibraryId;
```
Comment (brianwal) [inline app/src/main/java/com/example/captstone/models/Result.java:18]: final

File: app/src/main/java/com/example/captstone/models/Result.java [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    // Decodes array of result json results into business model objects
+    public static ArrayList<Result> fromJson(JSONArray jsonArray) {
+        ArrayList<Result> results = new ArrayList<>(jsonArray.length());
+        // Process each result in json array, decode and convert to business
+        // object
+        for (int i = 0; i < jsonArray.length(); i++) {
+            JSONObject resultJson;
+            try {
+                resultJson = jsonArray.getJSONObject(i);
+            } catch (Exception e) {
```
Comment (brianwal) [inline app/src/main/java/com/example/captstone/models/Result.java]: Not good practice to catch Exception.  Catch specific exceptions

**Comment by jackviel** [comment]
Added functionality to add media from reviews on the feed to a Netflix-like "My List". The goal is to make a gesture to easily add the media from the review to the user's list and have them show up in their profile.

Currently, there's a button on each review that shows up on the feed to add reviews to the user's list. When the button is pressed we get the review's object id since the review object contains all of the media's details, we then get the userMediaList Array from Parse and make an ArrayList as a local copy, then we check if the selected review object id isn't already in the ArrayList, if it's not then we add it and then put it back into the Parse database.

Since I'm saving review object ids, when I work on the profile page and displaying the contents of the user's "My List", I plan on checking if there are any duplicates in the list and delete remove them in case the user adds the same media from different reviews (super ducktapey for the moment).

Tested:
- Adding the same review to the user's list to make sure the same review object id isn't being saved more than once.

**Comment by jackviel** [comment]
Implemented a double-tap gesture so that the user can double tap a review anywhere and it'll add the media the review is about to the user's "my list".  

I set an on-touch-listener to the old button so that when it detects a double tap it'll call the old function that was implemented in the last commit to save the review object id to the current user's Parse object. I set the old button's alpha to 0 and made it cover the whole view for now, I'll make it so that it's when you double click the whole view instead of the invisible button.

I plan on adding something that pops up that shows that something was either added or not to the user's list, as it stands right now it isn't displaying anything.

Tested:
- That the function only goes off only when double-tapping
- That the function still works as it did the last time

**Comment by jackviel** [comment]
Implemented caching for the feed queries so that if there's no internet, cached reviews will still be shown in the feed.

Parse had a built-in attribute that could be changed to utilize cached API data so it was accomplished with a single line of code, to satisfy the complex feature requirement I'll have to either implement caching by scratch utilizing android's internal storage and also checking for internet connectivity manually.

Tested:
- Starting the app with Airplane Mode on.
- Turning on Airplane Mode while the app is still open.

**Comment by jackviel** [comment]
Implemented an SQLite database to cache feed data so the app can be used without internet access.

Added a function that checks for internet connectivity, if there's internet connectivity when populating the feed, the app will query for reviews normally and insert them to the SQLite cache database. If there's no internet access then it will get the recent items added to the SQLite database and add them to the adapter. 

Created a cache model for reviews different from the Parse model in order for it to work.

Tested:
- Starting the app with airplane mode turned on.
- Turning on airplane mode and refreshing the feed when the app was started with internet access.
- Turning off airplane mode and refreshing the feed with internet access.

**Comment by jackviel** [comment]
Implemented various Material Design components including cards for the review feed, shapeable image views for the profile pictures, and the app's theme. 

Worked on overall styling and screen organization of the app. 

Fixed lag on the review feed by populating the feed adapter first and then storing the reviews on the SQLite database afterwards, instead of doing it at the same time as they were added to the adapter, also fixed the keyboard focus on the review compose fragment.

 Also removed some imports on various files that weren't being used.

**Comment by jackviel** [comment]
Started implementing the trending fragment replacing the compose fragment.

Currently just shows the currently trending movies, songs, and books.

Had to create new client classes to get different data from the APIs, and had to incorporate a new API (NY Times API) to get the currently trending books (NY Best Seller's). 

Currently the trending fragment is a RecyclerView with the fetched trending media.

**Comment by jackviel** [comment]
Implemented a tab layout for the trending fragment with a ViewPager under it populated with the former trending feed fragment. 

Worked on overall styling of the app, changed style colors and the review feed items layout.

Fixed trending book title capitalization.

**Comment by jackviel** [comment]
Implemented the selected trending media item fragment that shows up when you click a certain title in the trending media fragment. 

The fragment consists of the media's details and all the reviews that have been written about it.

When clicking a trending title, its details get wrapped up in parcel and sent to the new fragment created, where it gets unwrapped and displayed, then to display the reviews about it I have a RecyclerView that gets populated by a ParseQuery, similar to the main review feed, but we check the review's mediaTitle attribute to see if the review should be populated.

To do:
- Work on styling
- Add a message when there's no reviews
- Add a way to prompt user to compose their reviews about the trending title

Tested:
- Clicking on a trending title without reviews
- Clicking on a trending title with reviews

**Comment by jackviel** [comment]
Implemented the 'My Reviews' tab in the profile fragment.

When the 'My Reviews' tab gets clicked, the tab layout populates a ViewPager with the MyReviewsFragment which consists of a RecyclerView being populated in a similar manner to the selected trending media fragment, which queries reviews and checks if the user who composed the review is the current user using the app.

Still need to work on styling, and should look to instead of doing a query on ALL reviews and checking if the user composed it, adding review ids to an array that is stored in the ParseDB under the user when a new review is composed.

Tested:
- Different users to check if the RecyclerView was being populated correctly

**Comment by jackviel** [comment]
Implemented the media search from the previous compose screen into the action bar that is seen in all fragments, worked on correctly displaying the RecyclerView only when needed.

The plan is to add the same functionality that's in the trending fragment where it sends you to the selected media's page when you click a media title and see its reviews, and then add functionality to add your own review.

Tested:
- Empty search queries.
- Queries where it'll get results from some API but not all.

**Comment by jackviel** [comment]
Implemented functionality to add reviews on the SelectedMediaActivity.

When a user searches for a title or selects a currently trending title the SelectedMediaActivity will show the movie details, its reviews, and a button to add a review about it.

Here's how the app is currently looking: https://i.imgur.com/SDOFfeg.mp4

Still have to work on styling and removing the submit review pop-up after submitting a review, then refreshing the RecyclerView.

Tested:
- Trying to post a review on media that already has reviews.
- Trying to post a review on media that doesn't have reviews.
- Not adding the required review body and trying to submit.

**Comment by jackviel** [comment]
Cleaned up various comments with code that wasn't used anymore. Also made log TAGs in various classes and myDatabase in the ParseApplication class private.

**Comment by jackviel** [comment]
Implemented rating bar functionality.

Added a reviewRating entry on the Review class in the Parse database, when submitting a review the user can select a rating to give the media title, when submitting the review the selected rating will be saved along with the rest of the review details.

Updated the saveReview function in the SelectedMediaActivity to include a float called reviewRating. Also updated the Review model class with a set and get function for the new reviewRating column in the Parse database.

Tested: 
- Submitting with a certain rating and making sure the same rating is being saved and displayed in the feed correctly.

**Comment by jackviel** [comment]
Worked on showing 'my list' contents in the profile tab, worked on showing the correct icon on reviews depending on if the media is on the user's list or not, and also added the functionality to click the ImageButton to add to 'my list' in addition to double-tapping the review.

When querying for reviews, it'll add each media title to a HashSet, then when displaying the review posts, it'll check if the review title is already in the HashSet, depending on the result the correct icon (either an add icon or a check icon) will be displayed through an the ImageButton.

When double-tapping the review or clicking the add-icon, the review objectId will be added into the userMediaList that's under the User class in the Parse database.

When going into the User's profile and clicking the 'My List' tab, it'll get the current user's userMediaList from the Parse database, then it will go through the review objectIds and add try to add the media title to a HashSet, if it wasn't already in the HashSet then we add the media to the RecyclerView to be displayed.

Tested: 
- Adding media to 'my list' that was already added from the same review.
- Adding media to 'my list' that was already added from a different review.
- Checked that the icon was being displayed correctly.
- Adding to 'my list' by clicking the ImageButton.
- Adding to 'my list' by double-tapping the Review.

**Comment by jackviel** [comment]
Implemented change username function.

When clicking on the change username button on the profile settings tab, the ChangeUsernameActivity will launch with an EditText and a submit button. When submitting a non-empty username, it will check if there's already a user named with the given username, if there is, an error toast will pop-up, if there isn't, it will change the user's username on the Parse database.

Tested:
- Changing username to an available username.
- Trying to change username to an already taken username.
- Making sure the posts update with the new username.
